### PR TITLE
Provide stricter typing for services and methods

### PR DESCRIPTION
**Pull request by thomaux**
This is based on the work of https://github.com/dwyl/aws-sdk-mock/pull/180, but seeing as that PR is most likely dead I opted to start a new one.

Main differences with #180:
- No merge conflicts with main
- Use tsd instead of dtslint for cleaner test setup

### CODE REVIEW

**Comment by nelsonic** [comment]
Hi @thomaux, 👋 
this PR looks _great_! 🥇 
I've invited you to be a contributor: https://github.com/dwyl/aws-sdk-mock/invitations 📥 
That way your commits will automatically run the CI/Tests without someone else having to trigger it. :shipit: 
And it means you can _assign_ PRs for review when you think they are ready. 👍

**Comment by thomaux** [comment]
Hey @nelsonic , thanks! I'm happy to help wherever possible 😄

File: package-lock.json [hunk 1]
```diff
@@ -3640,35 +4284,37 @@
...
-          }
-        },
         "punycode": {
           "version": "2.1.1",
           "bundled": true,
           "dev": true
         },
-        "react-is": {
-          "version": "16.13.1",
+        "react": {
```
Comment (nelsonic) [inline package-lock.json:4292]: seems odd to be including **`react`** as a dependency/devDependency here ... what is the usecase? 🤷‍♂️

Comment (nelsonic) [reply package-lock.json:4292]: Appears to be a `devDependency` of `tsd`: https://github.com/SamVerschueren/tsd/blob/5347e25e2ffb28739ea3e877faddc22d2e9e49b9/package.json#L62
Totally get that they might want to test their code against `react`. 
But then why does it have to appear in our dependency tree? 🙄 
When we `npm install` do we _need_ to have the grandchild `devDependencies` for our tests (`tsd`) to pass? 💭

**Review by nelsonic — APPROVED** [review]
@thomaux these stricter types will be very welcome by people using typescript. Thanks! 🎉 

Quick follow-up question: do you agree that these stricter types are a "**_breaking_ change**" from a `TypeScript` perspective?
(_just trying to decide if we need to publish a new "major" version of the NPM package or if a **minor** update e.g. **`5.5.0`** is appropriate ..._ 💭)

File: package-lock.json (hunk 1 above)
Comment (thomaux) [reply package-lock.json:4292]: Good catch! I wouldn't expect their devDependencies to show up in ours. I'm currently on holiday without my laptop (bold, I know) so I can have a look at this in 2 weeks

**Comment by thomaux** [comment]
> @thomaux these stricter types will be very welcome by people using typescript. Thanks! 🎉
> 
> Quick follow-up question: do you agree that these stricter types are a "**_breaking_ change**" from a `TypeScript` perspective? (_just trying to decide if we need to publish a new "major" version of the NPM package or if a **minor** update e.g. **`5.5.0`** is appropriate ..._ 💭)

These changes will break usage of the library when the user would call mock on unknown AWS services or methods. But then again, in those cases, the library wouldn't work either way. So I would think this change can be released as a minor version.

File: package-lock.json (hunk 1 above)
Comment (thomaux) [reply package-lock.json:4292]: Good catch, I wouldn't expect their devDependencies to show up in this tree. I'm currently on holiday without my laptop so I can't dig in deeper right now. After the 12th of December I'll have a look as to why this dependency is showing up here

**Comment by nelsonic** [comment]
@thomaux assigning this back to you for a quick `merge-conflict` fix then we can merge and ship it ASAP. Thanks. 👍

**Comment by thomaux** [comment]
> @thomaux assigning this back to you for a quick `merge-conflict` fix then we can merge and ship it ASAP. Thanks. 👍

I'll pick this up once I have access to my dev machine, will be somewhere after 12th of December. I'll also have a look at that point if we can avoid the react dependencies.

**Comment by thomaux** [comment]
@nelsonic brought the source branch up to speed with main, so the merge conflicts are resolved. Waited with this until after #256 was fixed.

**Review by nelsonic — APPROVED** [review]
@thomaux stricter typing looks great! Thanks! 🙌 
(apologies for delay in review/merge ... thanks for your patience!)

**Comment by thomaux** [comment]
Thanks @nelsonic and no worries!

**Pull request merged**